# Analyzing the dataset of used Cars from eBay Kleinanzeigen

# 1.


In [431]:
import pandas as pd
import numpy as np

In [432]:
autos = pd.read_csv("autos.csv", encoding = "Windows-1252")

In [433]:
autos


dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

# 2.

In [434]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [435]:
all_vehicle_type = autos["vehicle_type"].value_counts()
print(all_vehicle_type)

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64


Limousine has the most seller

In [436]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_PS              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

In [437]:
autos.loc[autos["vehicle_type"].isnull(),:]

date_crawled                                               name  \
8      2016-03-22 16:51:34                                         Seat_Arosa   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
46     2016-03-31 10:53:28                                       BMW_mit__Lpg   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
59     2016-03-17 17:50:54               Mercedes_A_Klasse_W_168__A_140_gruen   
64     2016-04-05 07:36:19                 Autotransport__Abschlepp_Schlepper   
65     2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
84     2016-03-27 19:52:54                                     Renault_twingo   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
92     2016-03-25 19:47:32                                       Zu_verkaufen   
101    2016-03-22 17:51:49    Schnaepchen_in_einem_Jahr_OLDTEIMER_KENNZEICHEN   
113    2016-04-03 14:58:29                              Golf_4_Anfaenger_auto   
135    2016-03-12 11:00:10  Opel_Meriva_B_Panoramadach__Sitz__und_Lenkradh...   
147    2016-03-15 14:52:55                          VW_Polo_Ersatzteilspender   
164    2016-03-13 20:39:16  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
196    2016-03-22 17:54:07              Bmw_316i_lpg_coupe___Tausch_moeglich_   
197    2016-04-05 10:36:24                              VW_Polo_9N_an_Bastler   
203    2016-03-29 21:47:01                                   BMW_316i_Touring   
227    2016-04-04 17:54:21                               Als_Bastler_Fahrzeug   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
253    2016-03-27 13:25:18             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017   
256    2016-04-03 20:50:38                     Passat_1.9TDI_4Motion_Highline   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
276    2016-03-09 16:58:42              Peugeot_106_1_4_55kw_Nuer_heute!!!!!!   
...                    ...                                                ...   
49689  2016-03-14 15:44:53                                      VW_GOLF_V_1.9   
49696  2016-03-08 13:36:21               Audi_A3_1_6_Attraction_mit_TÜV_10/17   
49698  2016-03-13 17:48:17                                      opel_vectra_b   
49713  2016-03-22 08:51:10                         Toyota_Carina_1600_GL_TA60   
49715  2016-03-10 22:54:49                                         a4_b5_1.8t   
49717  2016-03-08 16:50:11                                         Mazda_323C   
49722  2016-03-29 10:37:17                                       fghgfhfghfgh   
49725  2016-03-26 12:58:05                       VW_Passat_2.8l_V6_Syncro_AHK   
49728  2016-03-25 16:00:11                     Ford_Focus_1.4_16V_Bastlerauto   
49731  2016-03-28 13:40:31  Renault_MEGANE_1.6_16V__neu_ZANHRIEMEN_TOP_ZUS...   
49739  2016-03-17 15:47:34                          Reparaturfaelligen_twingo   
49750  2016-03-29 22:37:59                                 Volkswagen_Polo_6n   
49770  2016-03-15 12:54:26                     VW_Polo_6n_Tuev_Neu!__1.6_75PS   
49796  2016-03-09 09:38:38                        Opel_corsa_1.4_zu_verkaufen   
49819  2016-03-26 09:53:05  Audi_A3_1_8__Klima__5_Tuerig__Bremsen_neu__Alu...   
49841  2016-03-11 15:37:02                                  Passat_abzugeben.   
49844  2016-03-26 19:43:51                                Fahrzeug_Ankauf!!!!   
49845  2016-03-21 00:50:15  Schlachte_VW_Shar

In [438]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [439]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Change from camel case to snake case since python use snake case 

# 3.

In [440]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

CANDIDATES FOR DROPPING

In [441]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [442]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [443]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

END

Converting price and odometer to int dtype
Changing the column name with the unit of price and odometer

In [444]:
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(int))

In [445]:
print(autos["price"].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


In [446]:
autos["odometer"] = (autos["odometer"]
                  .str.replace("km","")
                  .str.replace(",","")
                  .astype(int))

In [447]:
#This are string values, need to convert to int
price_odometer = autos[["price","odometer"]]
print(price_odometer.head())

   price  odometer
0   5000    150000
1   8500    150000
2   8990     70000
3   4350     70000
4   1350    150000


In [448]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

# 4.
Finding ang outlier

In [449]:
print(autos["price_$"].unique().shape)
print(autos["price_$"].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64


Might be an outlier since the price is too large and i cant find the model of the car

In [450]:
autos[(autos["price_$"] == autos["price_$"].max())]


date_crawled                         name  seller offer_type  \
39705  2016-03-22 14:58:27  Tausch_gegen_gleichwertiges  privat    Angebot   

        price_$   abtest vehicle_type  registration_year    gearbox  power_PS  \
39705  99999999  control    limousine               1999  automatik       224   

          model  odometer_km  registration_month fuel_type          brand  \
39705  s_klasse       150000                   9    benzin  mercedes_benz   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
39705               NaN  2016-03-22 00:00:00               0        73525   

                 last_seen  
39705  2016-04-06 05:15:30

There are a lot of cars with no price

In [451]:
autos[(autos["price_$"] == 0)]

date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

In [452]:
autos["price_$"]

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price_$, Length: 50000, dtype: int64

In [453]:
autos["price_$"].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_$, dtype: int64

In [454]:
autos["price_$"].value_counts().sort_index().tail(20)

197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price_$, dtype: int64

In [455]:
autos["price_$"] = autos.loc[autos["price_$"].between(1,350000),"price_$"]

In [456]:
autos["price_$"].describe(include = "all")

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

REMOVED THE OUTLIER WITH THE VALUE 99999999 AND 0 PRICE. 
ITS SEEMS THAT 99999999 TO BE AN OUTLIER SINCE ITS ON EBAY. AS FOR THE ZERO PRICE. ITS AN ERROR

In [457]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [458]:
print(autos[(autos["odometer_km"] == autos["odometer_km"].max() )].shape)

(32424, 20)


There are no outliers in the odometer column since the max value is equal to the median and the 75 percentile.

# 5. 

In [459]:
autos["date_crawled"].value_counts(normalize = True,dropna=False)

2016-04-02 11:37:04    0.00006
2016-03-10 15:36:24    0.00006
2016-04-04 16:40:33    0.00006
2016-03-25 19:57:10    0.00006
2016-04-02 15:49:30    0.00006
2016-03-30 17:37:35    0.00006
2016-03-19 17:36:18    0.00006
2016-03-29 23:42:13    0.00006
2016-03-09 11:54:38    0.00006
2016-03-05 16:57:05    0.00006
2016-03-08 10:40:35    0.00006
2016-03-12 16:06:22    0.00006
2016-03-21 20:37:19    0.00006
2016-03-11 22:38:16    0.00006
2016-03-30 19:48:02    0.00006
2016-03-22 09:51:06    0.00006
2016-03-21 16:37:21    0.00006
2016-03-14 20:50:02    0.00006
2016-03-23 18:39:34    0.00006
2016-03-27 22:55:05    0.00006
2016-03-16 21:50:53    0.00006
2016-03-23 19:38:20    0.00006
2016-03-11 17:47:44    0.00004
2016-03-11 11:54:09    0.00004
2016-03-20 14:54:13    0.00004
2016-03-08 13:51:58    0.00004
2016-04-04 18:46:47    0.00004
2016-04-03 16:48:45    0.00004
2016-03-28 16:45:41    0.00004
2016-03-31 12:55:26    0.00004
                        ...   
2016-03-05 22:52:02    0.00002
2016-03-

In [460]:
autos["date_crawled"].describe()

count                   50000
unique                  48213
top       2016-04-02 11:37:04
freq                        3
Name: date_crawled, dtype: object

In [461]:
#autos_copy["date_crawled"]= autos_copy["date_crawled"].str.split("-")

In [462]:
autos["date_crawled"]= (autos["date_crawled"]
                       .str.split()
                     .str[0]
                    )

In [464]:
autos["date_crawled"].value_counts(dropna = False).sort_index()

2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1665
2016-03-09    1661
2016-03-10    1606
2016-03-11    1624
2016-03-12    1839
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1576
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1876
2016-03-22    1647
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1709
2016-03-30    1681
2016-03-31    1596
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1826
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

In [463]:
autos["date_crawled"].value_counts()

2016-04-03    1934
2016-03-20    1891
2016-03-21    1876
2016-03-12    1839
2016-03-14    1831
2016-04-04    1826
2016-03-07    1798
2016-04-02    1770
2016-03-19    1745
2016-03-28    1742
2016-03-29    1709
2016-03-15    1699
2016-04-01    1690
2016-03-30    1681
2016-03-08    1665
2016-03-09    1661
2016-03-22    1647
2016-03-26    1624
2016-03-11    1624
2016-03-23    1619
2016-03-10    1606
2016-03-31    1596
2016-03-25    1587
2016-03-17    1576
2016-03-27    1552
2016-03-16    1475
2016-03-24    1455
2016-03-05    1269
2016-03-13     778
2016-03-06     697
2016-04-05     655
2016-03-18     653
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

1. 2016-04-03 has the most date_crawled = 1934.
2. 2016-04-07 has the least date_crawled = 71
3. There are no values == 1.
4. We use Series.sort_index() method because Series.value_counts() returns a series of values of each date and date as its index. 
5. We used normalize to scale the column to 0-1 only
6. The first to crawl is in 2016-03-05 which has a value of 1269
7. The last to crawl is in 2016-04-07 which has a value of 71. Also the least date_crawled.

In [465]:
autos["ad_created"]= (autos["ad_created"]
                       .str.split()
                     .str[0]
                    )

In [466]:
autos["ad_created"].value_counts(dropna = False).sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       2
2016-02-11       1
2016-02-12       3
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
              ... 
2016-03-09    1662
2016-03-10    1593
2016-03-11    1639
2016-03-12    1831
2016-03-13     846
2016-03-14    1761
2016-03-15    1687
2016-03-16    1500
2016-03-17    1560
2016-03-18     686
2016-03-19    1692
2016-03-20    1893
2016-03-21    1886
2016-03-22    1640
2016-03-23    1609
2016-03-24    1454
2016-03-25    1594
2016-03-26    1628
2016-03-27    1545
2016-03-28    1748
2016-03-29    1707
2016-03-30  

In [467]:
autos["ad_created"].value_counts()

2016-04-03    1946
2016-03-20    1893
2016-03-21    1886
2016-04-04    1844
2016-03-12    1831
2016-03-14    1761
2016-04-02    1754
2016-03-28    1748
2016-03-07    1737
2016-03-29    1707
2016-03-19    1692
2016-04-01    1690
2016-03-15    1687
2016-03-30    1672
2016-03-08    1667
2016-03-09    1662
2016-03-22    1640
2016-03-11    1639
2016-03-26    1628
2016-03-23    1609
2016-03-31    1596
2016-03-25    1594
2016-03-10    1593
2016-03-17    1560
2016-03-27    1545
2016-03-16    1500
2016-03-24    1454
2016-03-05    1152
2016-03-13     846
2016-03-06     756
              ... 
2016-02-21       3
2016-02-14       2
2016-02-24       2
2016-02-18       2
2016-02-02       2
2016-02-20       2
2016-02-26       2
2016-02-09       2
2016-01-10       2
2016-02-05       2
2016-01-07       1
2016-01-22       1
2016-02-17       1
2016-02-16       1
2016-02-07       1
2016-01-03       1
2016-02-01       1
2015-08-10       1
2016-02-22       1
2016-01-16       1
2015-11-10       1
2016-02-08  

1. 2016-04-03 has the most ad created date with a value of 1946.
2. There are a lot of dates which has only one ad was created.
3. 2015-06-11 is the date in which the first ad was created
4. 2016-04-07 is the last date which have 64 ad created.

In [468]:
autos["last_seen"]= (autos["last_seen"]
                    .str.split()
                     .str[0]
                    )

In [469]:
autos["last_seen"].value_counts(normalize = True ,dropna = False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [470]:
autos["last_seen"].value_counts(dropna = False)

2016-04-06    11050
2016-04-07     6546
2016-04-05     6214
2016-03-17     1396
2016-04-03     1268
2016-04-02     1245
2016-03-30     1242
2016-04-04     1231
2016-03-31     1192
2016-03-12     1191
2016-04-01     1155
2016-03-29     1117
2016-03-22     1079
2016-03-28     1043
2016-03-21     1037
2016-03-20     1035
2016-03-24      978
2016-03-25      960
2016-03-23      929
2016-03-26      848
2016-03-16      822
2016-03-27      801
2016-03-15      794
2016-03-19      787
2016-03-14      640
2016-03-11      626
2016-03-10      538
2016-03-09      493
2016-03-13      449
2016-03-08      380
2016-03-18      371
2016-03-07      268
2016-03-06      221
2016-03-05       54
Name: last_seen, dtype: int64

1. 2016-04-06 has the greatest value in which an ad was seen online = 11050
2. 2016-03-05 has the least value in which an ad was seen online =54

In [471]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The max and min value is an outlier

In [472]:
autos["registration_year"] = autos.loc[(autos["registration_year"]
                                                  .between(autos["registration_year"].min()+900,autos["registration_year"].max()-7899)),"registration_year"]

# 6.

In [473]:
autos["registration_year"].describe()

count    49976.000000
mean      2003.366836
std          7.693943
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

I cleaned it by selecting only values between min(which is 1000) + 900 and max(which is 9999) - 7899
    #between 1900 and 2100

In [474]:
autos["registration_year"].value_counts(normalize = True)

2000.0    0.067112
2005.0    0.060329
1999.0    0.060029
2004.0    0.054766
2003.0    0.054566
2006.0    0.054186
2001.0    0.054086
2002.0    0.050684
1998.0    0.049084
2007.0    0.046102
2008.0    0.044641
2009.0    0.041980
1997.0    0.040579
2011.0    0.032696
2010.0    0.031955
2017.0    0.029074
1996.0    0.028894
2012.0    0.026473
2016.0    0.026333
1995.0    0.026273
2013.0    0.016128
2014.0    0.013326
1994.0    0.013206
2018.0    0.009845
1993.0    0.008904
2015.0    0.007984
1990.0    0.007904
1992.0    0.007824
1991.0    0.007123
1989.0    0.003622
            ...   
1966.0    0.000440
1975.0    0.000380
1969.0    0.000380
1965.0    0.000340
1964.0    0.000240
1963.0    0.000180
1910.0    0.000180
1959.0    0.000140
1961.0    0.000120
1956.0    0.000100
1962.0    0.000080
1958.0    0.000080
1937.0    0.000080
2019.0    0.000060
1950.0    0.000060
1934.0    0.000040
1941.0    0.000040
1951.0    0.000040
1954.0    0.000040
1955.0    0.000040
1957.0    0.000040
1952.0    0.

In [422]:
autos["brand"].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

# 7.

## Exploring the brand column

Choosing what brand to aggregate. Because we're computing the mean. And if data is not large enough. Then it wont be reliable

In [476]:
brands = autos["brand"].value_counts(normalize = True)

In [496]:
all_brand = {}
common_brands = brands[brands > 5].index

In [497]:
autos["brand"].value_counts(normalize = True).head()

volkswagen       0.21374
opel             0.10922
bmw              0.10858
mercedes_benz    0.09468
audi             0.08566
Name: brand, dtype: float64

In [498]:
for brand in all_brand_unique:
    brand_mean = autos.loc[(autos["brand"] == brand),"price_$"].mean()
    all_brand[brand] = int(brand_mean)

In [499]:
all_brand 

{'audi': 9212,
 'bmw': 8261,
 'ford': 3728,
 'mercedes_benz': 8536,
 'opel': 2944,
 'volkswagen': 5332}

In [501]:
all_brand_mean = pd.Series(all_brand)
print(all_brand_mean)

audi             9212
bmw              8261
ford             3728
mercedes_benz    8536
opel             2944
volkswagen       5332
dtype: int64


In [ ]:
df = pd.DataFrame(all_brand_mean, columns = ["mean_price"])
print()